### DISCUSSION TITLE
Support for 2316 ROMs

### DISCUSSION THREAD

**Discussion by TomNisbet** [TomNisbet/TommyPROM#68] [Ideas]
@jscrane - I've updated the PromDevice23 code to detect the configuration of the CS lines on 2316 static ROM chips.  Can you give the new code a try?

After checking it out, you will need to enable the PROM_IS_23 code in configure.h. 

Before trying to read the chip, run the Unlock command.  This will cycle thru the different combinations of the CS lines to find one that successfully reads data.  If this command is successful, you can use the other commands, like Read or Dump, to access the data.

The chip you used to write the original code should be detected at all three CS lines HIGH.

Note that I swapped the CS1 and CS2 pin definitions to match other chips with this footprint, like the 2716.

**Comment by jscrane** [comment]
That seems to work fine:

```
 u
Writing the unlock code to disable Software Write Protect mode: just kidding!

Scanning Chip Select combinations starting at address 0000
CS3:L  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:H - 0b f8 0b 0b 0a 0a 0a 0d 20 20 0b 8a 48 98 48 a5 
CS3:H  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:H - 0b f8 0b 0b 0a 0a 0a 0d 20 20 0b 8a 48 98 48 a5 
Setting CS bits to CS3:H  CS2:H  CS1:H
OK
```

BTW I took the naming of the pins from [here](http://www.elektronikjk.com/elementy_czynne/IC/2316E.pdf).

Also the call to `delayMicroseconds()` at line 175 isn't needed; I just had that in there for testing.

**Comment by TomNisbet** [comment]
It is fairly standard for PROMs to have D3..D7, CE, A10, OE numbered from the bottom right of the chip and working up, so I just changed C2 and C1 so they are driven by the Arduino pins that usually drive the CE and OE signals.

It's interesting that your chip read successfully for both values of CS3.  Did you have all three CS pins wired up?  I got the same kind of result in my test, but I was using a 28C256 with CE and OE, so it effectively only has two chip selects.

Does your board have more than one ROM?  I'd expect that you would see different CS results for each one.

It would be great to start building up a database of chips in the documentation that says, "the Apple II Basic ROM 1 uses HLH for the chip select values."  I've seen conflicting posts on various boards where people are trying to read 2316s and it isn't clear that different chips will need different CS settings.  I could see a dedicated documentation page for this chip to cover the programmable chip selects.

**Reply by jscrane** [reply]
There are 5 ROMs: the monitor and 4 basic-in-rom ([Microsoft](https://en.wikipedia.org/wiki/Microsoft_BASIC)), all are 2kB from various months in 1981. The machine is a [UK101](https://en.wikipedia.org/wiki/Compukit_UK101).

I believe the Character Generator ROM is the same type but I can't remove that from the board right now. Interestingly, it is enabled all the time, by tying CS1-3 to +5v.

Monitor: `MONUK02` `8110`
```
Scanning Chip Select combinations starting at address 0000
CS3:L  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:H - 0b f8 0b 0b 0a 0a 0a 0d 20 20 0b 8a 48 98 48 a5 
CS3:H  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:H - 0b f8 0b 0b 0a 0a 0a 0d 20 20 0b 8a 48 98 48 a5 
```

Basic-1: `BASUK01` `8104`
```
Scanning Chip Select combinations starting at address 0000
CS3:L  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:H - 39 a6 55 a5 3f aa 0b a7 22 a9 00 ad 4e a9 b8 a7 
CS3:H  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:H - 39 a6 55 a5 3f aa 0b a7 22 a9 00 ad 4e a9 b8 a7 
```

Basic-2: `BASUK02` `8103`
```
Scanning Chip Select combinations starting at address 0000
CS3:L  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:H - 9c b0 a5 9e a4 9f 85 b8 84 b9 20 8a b2 a9 ac a0 
CS3:H  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:H - 9c b0 a5 9e a4 9f 85 b8 84 b9 20 8a b2 a9 ac a0 
```

Basic-3: `BASUK03-2` `8107`
```
Scanning Chip Select combinations starting at address 0000
CS3:L  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:H - c4 48 a5 c3 48 20 0c a7 4c 7a b0 a9 9e 20 03 ac 
CS3:H  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:H - c4 48 a5 c3 48 20 0c a7 4c 7a b0 a9 9e 20 03 ac 
```
There were 2 revisions of this chip (due to a bug in the string garbage collector).

Basic-4: `BASUK04` `8104`
```
Scanning Chip Select combinations starting at address 0000
CS3:L  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:L  CS2:H  CS1:H - c8 aa f0 c6 b1 73 45 b0 30 c4 e4 ac d0 1a b1 73 
CS3:H  CS2:L  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:L  CS1:H - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:L - ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff 
CS3:H  CS2:H  CS1:H - c8 aa f0 c6 b1 73 45 b0 30 c4 e4 ac d0 1a b1 73 
```

**Reply by TomNisbet** [reply]
I'm surprised that all of those chips are ignoring CS3 and are all using CS1 and CS2 as active high.  I would have expected that each of the 4 Basic chips would use a different combination of CS values so that they could be individually addressed without external chip select circuitry. 

Do the Basic ROMs have the CS2 and CS1 lines connected to each other or are they driven by another chip?  Is the CS3 just tied high or low?  The data sheet doesn't indicate that a CS line can be programmed for don't care.

That makes me start to doubt if my code is working correctly.

**Reply by jscrane** [reply]
Luckily computers from that era usually had schematics available!

The Monitor ROM has MCS (active high) connected to pin 18 and Phi2 (processor clock) connected to pins 20 and 21.

The Basic ROMs have BS0, BS1, BS2 and BS3 (active high) connected to pin 20, Phi2 connected to pin 21 and 18 tied to +5v. The select lines come from a 74LS139 decoder.

As I said above the Character Generator has all 3 select pins connected to +5v.

**Comment by TomNisbet** [comment]
The UK101 looks like an interesting system.  The first microcomputers I used were Ohio Scientific Challengers hooked up to teletype machines.  My school later moved up to newer OSI machines with integrated video and cassette tape storage, which was quite the upgrade from paper tape.  I learned Basic and 6502 assembly on those.

The first computer I built was a [Netronics Explorer/85](https://tomnisbet.github.io/Simple8085/docs/design) which was an 8085 single board system.

**Reply by jscrane** [reply]
The "integrated video and cassette" OSI machines you refer to could have been [these](https://en.wikipedia.org/wiki/Ohio_Scientific#Superboard_II,_Challenger_1P_(1978))?

The UK101 was a very simple machine: polled keyboard and cassette I/O, character "graphics", no interrupts. Ideal for learning Basic and 6502 assembly. The 6502 was also an ideal learning processor. My subsequent engineering studies and career in software were a direct result of buying that kit.

Over the years I have made an Arduino-based emulator for it and other old machines: [UK101](https://github.com/jscrane/UK101), [PET](https://github.com/jscrane/PET), [Apple-1](https://github.com/jscrane/Apple1), etc.

Recently I discovered a "modern" [spin](http://cbmsteve.ca/microuk101/index.html) on it and deciding to make one led me to get involved in TommyPROM --- thanks for that btw!

That Netronics looks interesting. I hadn't heard of those before.

**Comment by TomNisbet** [comment]
@jscrane - I've updated the [PROM-specific code documentation](https://tomnisbet.github.io/TommyPROM/docs/prom-specific-code#promdevice23) to include a description of the PromDevice23 module.  

There is also a [new page dedicated to the 2316](https://tomnisbet.github.io/TommyPROM/docs/2316-roms) that is still a work-in-progress.  This one isn't linked into the menus yet.  I'm hoping that this new page might help clear up some of the confusion that surrounds these chips.  Once a better version is in place, I may post on some of the vintage forums to let people know there is a tool for reading these chips.

I welcome any comments you have on the documentation.

**Reply by jscrane** [reply]
That page looks pretty good. I would group the UK101 and OSI together because of their close relationship.

Commodore spelled it **kernal** for some reason btw.

Maybe mention the TommyPROM32? I have found that much easier than fiddling around with breadboards.

Sorry I don't have any more old machines to help with the ROM archaeology! [6502.org](http://6502.org) or [Stardot](http://stardot.org.uk) would be good places to ask though.
